### PR TITLE
Add a timeout (default: 1 second)

### LINKED ISSUE #36: Proposal to make use of `redisConnectWithTimeout`
We currently use `redisConnect` to connect to the Redis server. If hiredis cannot connect to the backend, it will timeout at some point but with some disadvantages compared to using `redisConnectWithTimeout` as there's no sensible default, and developers might want to tune that value (building [holycorn](https://github.com/franckverrot/holycorn)'s Redis data wrapper, I would like to give an opportunity to the users to override the defaults).

Asking the `mruby-redis` users, would you like that type of feature?

Thanks in advance!

### PR DESCRIPTION
**Pull request by franckverrot**
This closes #36.

### CODE REVIEW

**Comment by matsumoto-r** [comment]
Hmm, code looks good to me. But, test failed with segfault.  Do you know this cause?

**Comment by franckverrot** [comment]
Yeah I'm not sure what to do. It's building on both my OSX setup **and** my docker container (ubuntu 14.04).

**Comment by matsumoto-r** [comment]
We test mruby-redis per commit on Travis CI. 

https://travis-ci.org/matsumoto-r/mruby-redis/builds/71160666

On travis ci, this PR was failed with segfault. And, I ran the test again, but still was failed. Could you fix this problem?

**Comment by franckverrot** [comment]
I indeed checked the build and even rebased to change the SHA1: same issue.

I have no idea right now as even a brand new Linux container is running the
build without any problem :-(

Did you try running the test locally too (on your machine)?

**Comment by franckverrot** [comment]
Got it. Downgraded from 14.04 to 12.04 and got the seffault! Glad and sad at the same time :sob:

**Comment by franckverrot** [comment]
So I think there's a [bug here](https://github.com/mruby/mruby/blob/master/src/class.c#L766):

```c
    // mrb_get_args
    case '?':
      {
        mrb_bool *p;

        p = va_arg(ap, mrb_bool*);
        *p = given; // <---- HERE
      }
      break;
```

Changing 

```c
        *p = given;
```
to
```c
        p = &given;
```
fixes the problem. What do you think?

I'll discuss this on mruby's issue tracker but I'm not sure they support 12.04 as I'm having a lot of warnings/test failures when running `mruby-redis` on 12.04. :disappointed:

File: src/mrb_redis.c [hunk 1]
```diff
@@ -61,9 +64,11 @@ mrb_value mrb_redis_connect(mrb_state *mrb, mrb_value self)
     DATA_TYPE(self) = &redisContext_type;
     DATA_PTR(self) = NULL;
 
-    mrb_get_args(mrb, "oo", &host, &port);
+    if(mrb_get_args(mrb, "oo|i?", &host, &port, &timeout) == 3) {
+      timeout_struct.tv_sec = timeout;
+    }
```
Comment (matsumoto-r) [inline src/mrb_redis.c]:
```
    mrb_get_args(mrb, "oo|i", &host, &port, &timeout);
    timeout_struct.tv_sec = timeout;
```

Do `mrb_get_args` need `?` format?

**Comment by matsumoto-r** [comment]
mruby-redis didn't occur that segfault by test, but test failed.

https://travis-ci.org/matsumoto-r/mruby-redis/builds/71377377

Could you modify test code to work correctly?

**Comment by franckverrot** [comment]
Nah sorry, this last commit doesn't fix anything (I worked on it last night).

So no, I have no clue about what's going on on Travis :-(

**Comment by franckverrot** [comment]
Do you think that upgrading to the latest Travis container-based infrastructure would be OK?

**Comment by matsumoto-r** [comment]
OK, so could you modify `travis.yml` which supports the latest Travis container-based infrastructure?

**Comment by franckverrot** [comment]
@matsumoto-r Done. And the tests finally pass. Thanks!

**Comment by matsumoto-r** [comment]
Wow! It's great work! Merged to master.

**Comment by franckverrot** [comment]
Thanks, really appreciate it :+1:

**Pull request merged**